openthread: Move OpenThread implementation from net to modules

ArekBalysNordic opened the pull request:
Move OpenThread-related code from `zephyr/subsys/net/l2/openthread/openthread.c` to `zephyr/modules/openthread/platform/openthread.c`.

The primary goal of this refactor is to enable the use of OpenThread as an independent module, without the necessity of Zephyr's networking layer.

This change is particularly beneficial for simple applications that have their own implementation of the IEEE802.15.4 driver and do not require a networking layer. These applications can now disable Zephyr's L2 and IEEE802.15.4 shim layers and directly use the OpenThread module, saving valuable kilobytes of memory.

In this approach, if the `CONFIG_NET_L2_OPENTHREAD` Kconfig option is set, Zephyr's L2 and IEEE802.15.4 layers will be used, and everything will function as before. The main difference is that the Zephyr's L2 layer now uses the OpenThread module, no longer implementing it.

While most of the functions in `include/net/openthread.h` have been deprecated, they are still available for use to maintain backwards compatibility.

ArekBalysNordic committed:
openthread: Move OpenThread implementation from net to modules

Move OpenThread-related code from
zephyr/subsys/net/l2/openthread/openthread.c to
zephyr/modules/openthread/platform/openthread.c.

The primary goal of this refactor is to enable the use
of OpenThread as an independent module, without the necessity
of Zephyr's networking layer.

This change is particularly beneficial for simple applications
that have their own implementation of the IEEE802.15.4 driver
and do not require a networking layer. These applications can
now disable Zephyr's L2 and IEEE802.15.4 shim layers and
directly use the OpenThread module, saving valuable kilobytes
of memory.

In this approach if the CONFIG_NET_L2_OPENTHREAD
Kconfig option is set, Zephyr's L2 and IEEE802.15.4 layers
will be used, and everything will function as before.
The main difference is the Zephyr's L2 layer now uses
the OpenThread module, no longer implementing it.

While most of the functions in include/net/openthread.h
have been deprecated, they are still available for use to
maintain backwards compatibility.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

ArekBalysNordic committed:
doc: releases: add release note for OpenThread update

Add an entry to the release note and migration guide regarding
moving the OpenThread implementation from the L2 networking layer
to the OpenThread module.

The migration guide note contains information about deprecated
functions, explains what functions should be used instead and
provides a guide for required updates in users' applications.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

ArekBalysNordic commented:
I will add a release note and migration guide entries to this PR soon.

pdgendt commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -429,248 +269,54 @@ int openthread_send(struct net_if *iface, struct net_pkt *pkt)
+	/* Check whether the OpenThread and its instance were properly initialized by the OpenThread
+	 * module
+	 */
+	__ASSERT(openthread_init(ot_receive_handler, (void *)ot_l2_context),
+		 "OT instance is not initialized");
```
Do not call functions inside of `__ASSERT` statements, as these could expand to nothing.

ArekBalysNordic replied:
I've removed it from there since it is checked in openthread.c in modules.

pdgendt commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -429,248 +269,55 @@ int openthread_send(struct net_if *iface, struct net_pkt *pkt)
+	/* To keep backward compatibility use the additional state change callback list from the ot
+	 * l2 context and register the callback to the openthread module.
+	 */
+	 sys_slist_init(&ot_l2_context->state_change_cbs);
+	 openthread_state_change_callback_register(&ot_l2_state_changed_cb); 
```
Nit

```suggestion
	sys_slist_init(&ot_l2_context->state_change_cbs);
	openthread_state_change_callback_register(&ot_l2_state_changed_cb); 
```

ArekBalysNordic replied:
Right, just spotted. Thanks.

rlubos commented on include/zephyr/net/openthread.h:
```diff
@@ -62,22 +65,40 @@ struct openthread_context {
+ * @param aMessage The message to receive.
+ * @param aContext The context to pass to the callback.
+ */
+typedef void (*openthread_receive_cb)(otMessage *aMessage, void *aContext);
```
nit: Param names should just be `message` and `context` for Zephyr code.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -429,248 +269,50 @@ int openthread_send(struct net_if *iface, struct net_pkt *pkt)
-		if (err != OT_ERROR_NONE) {
-			NET_ERR("Could not set state changed callback: %d", err);
-		}
+	openthread_mutex_lock();
```
Is this mutex lock/unlock needed here? We're not calling any OT APIs here.

ArekBalysNordic replied:
That's right, removed.

rlubos reviewed (commented):
Generally looks fine, but need to figure out what's broken - apart from the twister issues, BabbleSim shows the Thread network fails to start.

ArekBalysNordic commented:
> Generally looks fine, but need to figure out what's broken - apart from the twister issues, BabbleSim shows the Thread network fails to start.

Yeah, I'm debugging it locally to figure it out.

ArekBalysNordic committed:
drivers: nxp: Do not use private L2 data structure.

An implementation should use a dedicated function to get
openthread instance instead of using the private L2 data.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

ArekBalysNordic committed:
samples: Enable OpenThread logging in all related samples

Some OpenThread logs were moved from L2 to the OpenThread module,
so enable logging from there to keep previous logging visibility.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

rlubos commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -332,9 +332,101 @@ Networking
+
+    * Now use:
+
+      * ``openthread_mutex_lock``
```
Minor suggestion: for the new functions we could use func reference to generate links in the documentation 
```
:c:func:`openthread_mutex_lock`
```
 etc

ArekBalysNordic replied:
Right, I have it. All new things (structs, kconfigs, functions) are now as links. Deprecated functions remain as they were previously.

rlubos reviewed (dismissed):
LGTM, thanks for the detailed migration guide.

Damian-Nordic commented on include/zephyr/net/openthread.h:
```diff
@@ -62,25 +65,76 @@ struct openthread_context {
+ * callbacks as needed can be added as long as each of them
+ * are unique pointers of struct openthread_state_changed_cb.
+ *
+ * @note Such structure should not be allocated on stack.
```
```suggestion
 * @note You may destroy the object only after it is unregistered from the callback list.
```

Damian-Nordic commented on include/zephyr/net/openthread.h:
```diff
@@ -111,23 +165,44 @@ struct openthread_state_changed_cb {
+ *
+ * @param cb Callback struct to register.
+ */
+int openthread_state_change_callback_register(struct openthread_state_changed_callback *cb);
```
nit: The callback type is `...changed_callback` and the function is `...change_callback` (without `d`). Worth unifying :)

Damian-Nordic commented on include/zephyr/net/openthread.h:
```diff
@@ -159,9 +272,46 @@ struct otInstance *openthread_get_default_instance(void);
+ * @brief Set the additional callback for receiving packets.
+ *
+ * @details This callback is called once a packet is received and can be
+ * used to inject packets into the OpenThread stack.
```
"into the OpenThread stack" - did you mean "Zephyr networking stack"?

Damian-Nordic commented on include/zephyr/net/openthread.h:
```diff
@@ -159,9 +272,46 @@ struct otInstance *openthread_get_default_instance(void);
+int openthread_mutex_try_lock(void);
+
+/**
+ * @brief Unlock internal mutex after accessingOpenThread API.
```
```suggestion
 * @brief Unlock internal mutex after accessing OpenThread API.
```

Damian-Nordic commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,486 @@
+
+/* Global variables to store the OpenThread module context */
+static otInstance *openthread_instance;
+static struct k_mutex openthread_lock;
```
You could use `K_MUTEX_DEFINE` to initialize this statically. Same for `k_work` and `sys_slist_t`.

Damian-Nordic commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,486 @@
+		return -EINVAL;
+	}
+
+	/* This function must be called after openthread_init() */
```
If you statically initialize the syslist, this won't be needed.

Damian-Nordic commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -219,56 +100,51 @@ static void ot_state_changed_handler(uint32_t flags, void *context)
+static void ot_receive_handler(otMessage *message, void *context)
 {
-	struct openthread_context *ot_context = context;
+	struct openthread_context *ot_context = (struct openthread_context *)context;
```
nit: in C casts to/from `void*` are implicit so this was OK even before.

Damian-Nordic commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -429,248 +269,47 @@ int openthread_send(struct net_if *iface, struct net_pkt *pkt)
-					ot_receive_handler, ot_context);
-
-#if defined(CONFIG_OPENTHREAD_NAT64_TRANSLATOR)
+	openthread_init();
```
check return value?

Damian-Nordic reviewed (approved):
Looks great. Some mostly nits found.

ArekBalysNordic commented:
@Damian-Nordic, I've provided all suggested changes, thanks!

ArekBalysNordic commented:
Discussed offline that the new `openthread.c` file should be located outside the `platform` directory since it is more of a utility than a platform implementation. The same as `shell.c` that doesn't implement the OpenThread platform as well.

edmont reviewed (approved):
Looks great, thanks!
